### DISCUSSION TITLE
Role based authorization in combination with IAuthorizationFilter

### DISCUSSION THREAD

**Discussion by plombaer** [CoreWCF/CoreWCF#1350] [Q&A]
Hello,
I implemented role based authorization using bearer token.
I'd like to intercept whether the user is authorized or not by using interface IAuthorizationFilter or any other method.
My service :
```
    [Authorize(Roles = "Blabla")]
    public async Task<Response> TestAsync(Request request, [FromServices] Service service)
```
The IAuthorizationFilter implementation :
```
builder.Services.AddScoped<IAuthorizationFilter, AuthorizationLoggingFilter>();

    public sealed class AuthorizationLoggingFilter : IAuthorizationFilter
    {
        public void OnAuthorization(AuthorizationFilterContext context)
        {
            var isAuthorized = context.HttpContext.User.Identity?.IsAuthenticated ?? false;
        }
    }

```
Unfortunately the OnAuthorization method never gets called.
Do I need to also implement ServiceFilterAttribute, TypeFilterAttribute or something else ?

Thanks for your help

**Comment by g7ed6e** [comment]
CoreWCF do not pull IAuthorizationFilter from DI indeed. I think we should be able to implement this in ImmutableDispatchRuntime.ProcessMessageAsync or maybe in AuthorizationBehavior.AuthorizePolicyAsync. @mconnew what do you think ?

**Reply by mconnew** [reply]
My only concern is potentially breaking someones service. I don't currently know enough about the feature to make an informed decision about this. Could someone have an IAuthorizationFilter configured for say an MVC endpoint in their app and pulling this in now breaks their service? It's the "unintentionally depending on a behavior" case that usually bytes you with changes like this. Is a filter intended to change an authorization decision or just for adding more information or doing an ancillary action such as looks to be the case with AuthorizationLoggingFilter?

**Reply by g7ed6e** [reply]
@mconnew I understand your concern. 
To avoid the side effect you mentioned we should stick to an explicit way to apply the authorization logic to an operation using attribute. 

@plombaer  We already support the `AuthorizeAttribute` specifying the `Policy` property which can link to `Requirement` and pull `AuthorizationHandler<TRequirement>` from DI. Please see https://learn.microsoft.com/en-us/aspnet/core/security/authorization/policies?view=aspnetcore-8.0.
Having a second look to the code posted above it looks like it uses the `AuthorizeAttribute` specifying the `Roles` property which is not something CoreWCF support and have unit tests. Please see that blog post https://corewcf.github.io/blog/2023/02/19/aspnetcore-authorization-support

edit: you should receive a `COREWCF_0202` compilation warning
